resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -182,10 +185,25 @@ private[spark] class KubernetesClusterSchedulerBackend(
       } else if (totalExpectedExecutors.get() <= runningExecutorPods.size) {
         logDebug("Maximum allowed executor limit reached. Not scaling up further.")
       } else {
+        val executorPodsWithIPs = EXECUTOR_PODS_BY_IPS_LOCK.synchronized {
+            executorPodsByIPs.values.toList
+          }
+        val nodeToLocalTaskCount = mutable.Map[String, Int]() ++
+          KubernetesClusterSchedulerBackend.this.synchronized {
+            hostToLocalTaskCount
+          }
+        for (pod <- executorPodsWithIPs) {
+          // Remove cluster nodes that are running our executors already.
+          nodeToLocalTaskCount.remove(pod.getSpec.getNodeName).nonEmpty ||
+            nodeToLocalTaskCount.remove(pod.getStatus.getHostIP).nonEmpty ||
+            nodeToLocalTaskCount.remove(
+              InetAddress.getByName(pod.getStatus.getHostIP).getCanonicalHostName).nonEmpty
+        }
+        val preferredNodes = nodeToLocalTaskCount.toMap[String, Int]
```

Comment:
can you pull this chunk of code out into a method named something like `currentNodesWithTaskCounts`

Also don't like the name `preferredNodes` here -- this doesn't contain the preferred nodes for the new executor, it's actually a list of nodes that don't currently have executors along with their task count. 
 Which is later converted into a prioritization